authorizations/and/src/contract.rs

```diff
@@ -0,0 +1,128 @@
+
+const UPDATE_REPLY_ID: u64 = 1_000_000;
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    PARENT.save(deps.storage, &msg.parent)?;
+    for child in msg.children {
+        CHILDREN.save(deps.storage, child, &cosmwasm_std::Empty {})?;
+    }
+    Ok(Response::default().add_attribute("action", "instantiate"))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    if info.sender != PARENT.load(deps.storage)? {
+        return Err(AuthorizationError::Unauthorized {
+            reason: Some("Only the parent can execute on this contract".to_string()),
+        }
+        .into());
+    }
+
+    match msg {
+        ExecuteMsg::AddChild { addr } => {
+            CHILDREN.save(deps.storage, addr, &Empty {})?;
+            Ok(Response::default().add_attribute("action", "allow"))
+        }
+        ExecuteMsg::RemoveChild { addr } => {
+            CHILDREN.remove(deps.storage, addr);
+            Ok(Response::default().add_attribute("action", "remove"))
+        }
+        ExecuteMsg::UpdateExecutedAuthorizationState { msgs, sender } => {
+            // This authorization is a passthrough, so any update messages
+            let response = Response::default()
+                .add_attribute("action", "execute_authorize")
+                .add_attribute("authorized", "true");
+
+            CHILDREN
+                .range(deps.storage, None, None, Order::Ascending)
+                .fold(
+                    Ok(response),
+                    |acc, addr| -> Result<Response, ContractError> {
+                        // All errors from submessages are ignored since the validation should already have been done by the parent.
+                        // This assumes the parent handles the authorizations spec
+                        Ok(acc?.add_submessage(SubMsg::reply_on_error(
+                            wasm_execute(
+                                addr?.0.to_string(),
+                                &ExecuteMsg::UpdateExecutedAuthorizationState {
+                                    msgs: msgs.clone(),
+                                    sender: sender.clone(),
+                                },
+                                vec![],
+                            )?,
+                            UPDATE_REPLY_ID,
+                        )))
+                    },
+                )
+        }
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::Authorize { msgs, sender } => authorize_messages(deps, env, msgs, sender),
```

Comment:
Rename to `QueryMsg::IsAuthorized`?

`authorize_messages` sounds like an execute method when it's actually a query handler.